src/tactic/unify_equations.lean

```diff
@@ -0,0 +1,415 @@
+    },
+    pure (some next, ns)
+  else do
+    tgt ← target,
+    -- The following construction deals with a corner case involing
+    -- mutual/nested inductive types. For these, Lean does not generate
+    -- no-confusion principles. However, the regular inductive data type which a
+    -- mutual/nested inductive type is compiled to does have a no-confusion
+    -- principle which we can (usually? always?) use. To find it, we normalise
+    -- the constructor with `unfold_ginductive = tt`.
+    constructor_left ← get_app_fn_const_whnf lhs semireducible tt,
+    let no_confusion := constructor_left.get_prefix ++ "no_confusion",
+    pr ← mk_app no_confusion [tgt, lhs, rhs, h],
+    exact pr,
+    return (none, ns)
+
+/--
+Given `equ : C x₁ ... xₙ = D y₁ ... yₘ` with `C` and `D` constructors of the
+same datatype `I`:
+
+- If `C ≠ D`, we solve the goal by contradiction using the no-confusion rule.
+- If `C = D`, we clear `equ` and add equations `x₁ = y₁`, ..., `xₙ = yₙ`.
+-/
+meta def unify_constructor_headed : unification_step :=
+λ equ _ _ _ _ _ _ _,
+do {
+  (next, _) ← injection_with' equ [] `_ none,
+  try $ clear equ,
+  pure $
+    match next with
+    | none := goal_solved
+    | some next := simplified $ next.map expr.local_pp_name
+    end
+} <|>
+pure not_simplified
+
+/--
+For `type = I x₁ ... xₙ`, where `I` is an inductive type, `get_sizeof type`
+returns the constant `I.sizeof`. Fails if `type` is not of this form or if no
+such constant exists.
+-/
+meta def get_sizeof (type : expr) : tactic pexpr := do
+  n ← get_app_fn_const_whnf type semireducible ff,
+  resolve_name $ n ++ `sizeof
+
+lemma add_add_one_ne (n m : ℕ) : n + (m + 1) ≠ n :=
+begin
+  apply ne_of_gt,
+  apply nat.lt_add_of_pos_right,
+  apply nat.pos_of_ne_zero,
+  contradiction
+end
+-- Linarith could prove this, but I want to avoid that dependency.
+
+/--
+`match_n_plus_m n e` matches `e` of the form `nat.succ (... (nat.succ e')...)`.
+It returns `n` plus the number of `succ` constructors and `e'`. The matching is
+performed up to normalisation with transparency `md`.
+-/
+meta def match_n_plus_m (md) : ℕ → expr → tactic (ℕ × expr) :=
+λ n e, do
+  e ← whnf e md,
+  match e with
+  | `(nat.succ %%e) := match_n_plus_m (n + 1) e
+  | _ := pure (n, e)
+  end
+
+/--
+Given `equ : n + m = n` or `equ : n = n + m` with `n` and `m` natural numbers
+and `m` a nonzero literal, this tactic produces a proof of `false`. More
+precisely, the two sides of the equation must be of the form
+`nat.succ (... (nat.succ e)...)` with different numbers of `nat.succ`
+constructors. Matching is performed with transparency `md`.
+-/
+meta def contradict_n_eq_n_plus_m (md : transparency) (equ lhs rhs : expr) :
+  tactic expr := do
+  ⟨lhs_n, lhs_e⟩ ← match_n_plus_m md 0 lhs,
+  ⟨rhs_n, rhs_e⟩ ← match_n_plus_m md 0 rhs,
+  is_def_eq lhs_e rhs_e md <|> fail
+    "internal error in contradict_n_eq_n_plus_m:\nexpected {lhs_e} and {rhs_e} to be definitionally equal at transparency {md}.",
```

Comment:
Is this really an internal error, or just a case in which `unify_cyclic` doesn't apply?